File: frontend/src/types/card.ts
export interface Card {
  id: string;
  scryfall_id?: string;
  oracle_id?: string;
  name: string;
  mana_cost?: string;
  cmc?: number;
  type_line?: string;
  oracle_text?: string;
  power?: string;
  toughness?: string;
  colors?: string;
  color_identity?: string;
  rarity?: string;
  flavor_text?: string;
  artist?: string;
  collector_number?: string;
  image_uris?: Record<string, any>;
  prices?: Record<string, any>;
  legalities?: Record<string, any>;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
  set?: CardSet;
}

export interface CardSet {
  id: string;
  code: string;
  name: string;
  set_name?: string;
  release_date?: string;
  card_count?: number;
  icon_url?: string;
  metadata?: Record<string, any>;
}

export interface SearchFilters {
  colors?: string;
  set?: string;
  rarity?: string;
  type?: string;
}

export interface CardSearchResponse {
  data: Card[];
  meta: {
    total: number;
    page: number;
    per_page: number;
    total_pages: number;
    has_next: boolean;
    has_prev: boolean;
  };
}

export interface User {
  id: string;
  email: string;
  username: string;
  is_active: boolean;
  is_admin: boolean;
  preferences: Record<string, any>;
  created_at: string;
  last_login_at?: string;
}

export interface AuthResponse {
  access_token: string;
  refresh_token: string;
  token_type: string;
  expires_in: number;
}

export interface Collection {
  id: string;
  name: string;
  description: string;
  owner_id: string;
}

export interface CollectionCard {
  id: string;
  card: Card;
  quantity: number;
  condition?: string;
  is_foil?: boolean;
  added_at: string;
}

File: backend/app/schemas/card.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime, date
import uuid


class CardSetBase(BaseModel):
    code: str
    name: str
    release_date: Optional[date] = None
    card_count: Optional[int] = None
    icon_url: Optional[str] = None


class CardSetResponse(CardSetBase):
    id: uuid.UUID
    metadata: Dict[str, Any] = {}
    
    class Config:
        from_attributes = True


class CardBase(BaseModel):
    name: str
    mana_cost: Optional[str] = None
    type_line: Optional[str] = None
    oracle_text: Optional[str] = None
    power: Optional[str] = None
    toughness: Optional[str] = None
    colors: Optional[str] = None
    color_identity: Optional[str] = None
    rarity: Optional[str] = None
    flavor_text: Optional[str] = None
    artist: Optional[str] = None


class CardCreate(CardBase):
    scryfall_id: Optional[uuid.UUID] = None
    oracle_id: Optional[uuid.UUID] = None
    set_id: Optional[uuid.UUID] = None
    collector_number: Optional[str] = None
    image_uris: Optional[Dict[str, Any]] = None
    prices: Optional[Dict[str, Any]] = None
    legalities: Optional[Dict[str, Any]] = None
    metadata: Optional[Dict[str, Any]] = None


class CardResponse(CardBase):
    id: uuid.UUID
    scryfall_id: Optional[uuid.UUID] = None
    oracle_id: Optional[uuid.UUID] = None
    collector_number: Optional[str] = None
    image_uris: Optional[Dict[str, Any]] = None
    prices: Optional[Dict[str, Any]] = None
    legalities: Optional[Dict[str, Any]] = None
    metadata: Dict[str, Any] = Field(default_factory=dict, alias='extra_data')
    created_at: datetime
    updated_at: datetime
    set: Optional[CardSetResponse] = None
    
    class Config:
        from_attributes = True
        populate_by_name = True


class CardSearchParams(BaseModel):
    q: Optional[str] = None
    colors: Optional[str] = None
    set: Optional[str] = None
    rarity: Optional[str] = None
    type: Optional[str] = None
    page: int = 1
    per_page: int = 20


class CardSearchResponse(BaseModel):
    data: List[CardResponse]
    meta: Dict[str, Any]